resource-managers/mesos/src/main/scala/org/apache/spark/scheduler/cluster/mesos/MesosClusterScheduler.scala

```diff
@@ -782,16 +782,15 @@ private[spark] class MesosClusterScheduler(
     retryState: Option[MesosClusterRetryState], status: TaskStatus): MesosClusterRetryState = {
 
     val (retries, waitTimeSec) = retryState
-      .map { rs => (rs.retries + 1, rs.waitTime) }
+      .map { rs => (rs.retries + 1, if (isNodeDraining(status)) rs.waitTime else rs.waitTime * 2) }
       .getOrElse{ (1, 1) }
 
     // if a node is draining, the driver should be relaunched without backoff
     if (isNodeDraining(status)) {
       new MesosClusterRetryState(status, retries, new Date(), waitTimeSec)
     } else {
-      val newWaitTime = waitTimeSec * 2
-      val nextRetry = new Date(new Date().getTime + newWaitTime * 1000L)
-      new MesosClusterRetryState(status, retries, nextRetry, newWaitTime)
+      val nextRetry = new Date(new Date().getTime + waitTimeSec * 1000L)
+      new MesosClusterRetryState(status, retries, nextRetry, waitTimeSec)
```

Comment:
While this is a correct approach to solve the problem, this function still looks pretty complicated and contains two calls to `isNodeDraining`  which is suboptimal.  WDYT about refactoring it in this or similar way?

```
private def getNewRetryState(
  retryState: Option[MesosClusterRetryState], status: TaskStatus): MesosClusterRetryState = {
  retryState.map { rs =>
    // if a node is draining, the driver should be relaunched without backoff
    if (isNodeDraining(status)) {
      new MesosClusterRetryState(status, rs.retries + 1, new Date(), rs.waitTime)
    } else {
      val nextRetry = new Date(new Date().getTime + rs.waitTime * 1000L)
      new MesosClusterRetryState(status, rs.retries + 1, nextRetry, rs.waitTime * 2)
    }
  }.getOrElse {
    // this is the first retry which should happen without backoff
    new MesosClusterRetryState(status, 1, new Date(), 1)
  }
}
```

Reply:
I have a minor concern that this again changes original behavior a bit by making first retry delay zero instead of 1 second for not node draining cases.

To avoid the extra call, I'd rather simply save the value, like this `val noBackoff= isNodeDraining(status)` and use it. What do you say?

Reply:
My main point here is to get rid of this construct at all due to the newly discovered issue:
```
val (retries, waitTimeSec) = retryState
      .map { rs => (rs.retries + 1, rs.waitTime) }
      .map { rs => (rs.retries + 1, if (isNodeDraining(status)) rs.waitTime else rs.waitTime * 2) }
      .getOrElse{ (1, 1) }
```

I believe that JIT compiler will inline a call to this function so getting rid of double call it's not a matter of performance optimization but more of readability. I'm not insisting on the code piece posted above but IMO its readability is better because it can express the whole function as a nested conditional statement:
```
//pseudo-code
if (this is not the first retry) {
  if (draining) {
    return draining-specific retry
  } else {
    return retry state with applied backoff
  }
} else {
  return first retry state
}
```

As for zero delay, we can refactor the function even further and use `DateUtils` class from Apache Commons (which comes as Spark dependency) to manipulate the time increments:
```
val now = new Date()
retryState.map { rs =>
  val newRetries = rs.retries + 1
  // if a node is draining, the driver should be relaunched without backoff
  if (isNodeDraining(status)) {
    new MesosClusterRetryState(status, newRetries, now, rs.waitTime)
  } else {
    new MesosClusterRetryState(status, newRetries, DateUtils.addSeconds(now, rs.waitTime), rs.waitTime * 2)
  }
}.getOrElse {
  // this is the first retry which should happen without backoff
  new MesosClusterRetryState(status, 1, DateUtils.addSeconds(now, 1), 1)
}
```